core/src/main/java/org/elasticsearch/common/settings/Setting.java

```diff
@@ -574,16 +581,26 @@ public String getNamespace(Setting<T> concreteSetting) {
         }
     }
 
+    @FunctionalInterface
+    public interface Validator<T> {
+
+        void validate(T value, Map<Setting<T>, T> settings);
+
+        default Iterator<Setting<T>> settings() {
```

Comment:
Right now this carries the assumption that the settings that we validate against have the same type as the setting itself. I think this is okay for now, we can revisit later only if we need to?